core/src/cluster_slots.rs

```diff
@@ -0,0 +1,267 @@
+            let slots = epoch_slots.to_slots(root);
+            for slot in &slots {
+                if *slot <= root {
+                    continue;
+                }
+                let pubkey = Rc::new(epoch_slots.from);
+                if self.keys.get(&pubkey).is_none() {
+                    self.keys.insert(pubkey.clone());
+                }
+                let from = self.keys.get(&pubkey).unwrap();
+                let balance = self.default_peers.get(from).cloned().unwrap_or(0);
+                if self.self_id != **from {
+                    debug!(
+                        "CLUSTER_SLLOTS: {}: insert {} {} {}",
+                        self.self_id, from, *slot, balance
+                    );
+                }
+                self.cluster_slots
+                    .entry(*slot)
+                    .or_insert_with(HashMap::default)
+                    .insert(from.clone(), balance + 1);
+            }
+        }
+        self.cluster_slots.retain(|x, _| *x > root);
+        self.keys.retain(|x| Rc::strong_count(x) > 1);
+        self.since = since;
+    }
+    pub fn stats(&self) -> (usize, usize, f64) {
+        let my_slots = self.collect(&self.self_id);
+        let others: HashMap<_, _> = self
+            .cluster_slots
+            .iter()
+            .filter(|(x, _)| !my_slots.contains(x))
+            .flat_map(|(_, x)| x.iter())
+            .collect();
+        let other_slots: Vec<Slot> = self
+            .cluster_slots
+            .iter()
+            .filter(|(x, _)| !my_slots.contains(x))
+            .map(|(x, _)| *x)
+            .collect();
+
+        let weight: u64 = others.values().map(|x| **x).sum();
+        let keys: Vec<Rc<Pubkey>> = others.keys().copied().cloned().collect();
+        let total: u64 = self.default_peers.values().copied().sum::<u64>() + 1u64;
+        if !other_slots.is_empty() {
+            debug!(
+                "{}: CLUSTER_SLOTS STATS {} {:?} {:?}",
+                self.self_id,
+                weight as f64 / total as f64,
+                keys,
+                other_slots
+            );
+        }
+        (
+            my_slots.len(),
+            self.cluster_slots.len(),
+            weight as f64 / total as f64,
+        )
+    }
+    pub fn collect(&self, id: &Pubkey) -> HashSet<Slot> {
+        self.cluster_slots
+            .iter()
+            .filter(|(_, keys)| keys.get(id).is_some())
+            .map(|(slot, _)| slot)
+            .cloned()
+            .collect()
+    }
+
+    pub fn update_peers(
+        &mut self,
+        cluster_info: &RwLock<ClusterInfo>,
+        bank_forks: &RwLock<BankForks>,
+    ) {
+        let root = bank_forks.read().unwrap().root();
+        let (epoch, _) = bank_forks
+            .read()
+            .unwrap()
+            .working_bank()
+            .get_epoch_and_slot_index(root);
+        if self.epoch.is_none() || Some(epoch) != self.epoch {
+            let stakes = staking_utils::staked_nodes_at_epoch(
```

Comment:
I think this will give a mapping from delegates -> stake, based on: 
https://github.com/solana-labs/solana/blob/master/ledger/src/staking_utils.rs#L45-L48

is that what we want here? I'm not sure how delegates are represented ini gossip

Reply:
Yea, not sure.  What do you recommend?

Reply:
hmmm I don't have a great handle on delegation, but on second look it might be ok. Behind the scenes it seems to just sum up all the vote accounts for a given node pubkey: https://github.com/solana-labs/solana/blob/master/ledger/src/staking_utils.rs#L58, I guess that's how we indicate a delegate? 

If it's any reassurance, It seems that retransmit is also using this `staked_nodes_at_epoch`: https://github.com/solana-labs/solana/blob/master/core/src/retransmit_stage.rs#L63-L68.

Reply:
Ok, I think its fine then